dolosse/generic_json_consumer.py

```diff
@@ -0,0 +1,153 @@
+                valid = False
+        if isinstance(decode[other_key], dict):
+            populate(variables,
+                     decode[other_key],
+                     count,
+                     name+": ")
+        else:
+            if name in variables:
+                if valid:
+                    variables[name].append(following)
+                else:
+                    variables[name].append(
+                        variables[name][-1])
+            else:
+                variables[name] = [0]*(count-1)
+                if valid:
+                    variables[name].append(following)
+                else:
+                    variables[name].append(0)
+                    #Remember, the list might still be empty here
+
+
+def get_args():
+    """Creates and returns a kafka consumer"""
+    parser = ArgumentParser()
+    parser.add_argument("--topic", default="default",
+                        help="The name of the kafka topic from which to read")
+    parser.add_argument("--group", default="my-group",
+                        help="The kafka group in which the topic exists")
+    parser.add_argument("--server", default="localhost:9092",
+                        help="Name and port of the kafka server (e.g. localhost:9092)")
+    arguments = parser.parse_args()
+
+    # To consume latest messages and auto-commit offsets
+    return arguments.topic, arguments.group, arguments.server
+
+def redraw(figures):
+    """Redraw all figures from a dict of figures"""
+    for key in figures:
+        figures[key].canvas.draw()
+        figures[key].canvas.flush_events()
+
+
+def update_data(data, value):
+    """Update the list of data, given some json input"""
+    try:
+        update = loads(value)
+    except decoder.JSONDecodeError as problem:
+        print("Error: Flawed JSON in value: ",
+              value)
+        print("Error message: ", problem)
+        update = ""
+    data.update(update)
+
+
+def graph(topic, group, server):
+    """Main function - reads json data from a kafka topic and graphs it"""
+
+    screen = initscr()
+    cbreak()
+    screen.nodelay(True)
+    screen.getch()
+
+    print("Generic Graphing Consumer")
+    print("Press any key to exit...")
+
+    data = {}
+    line = {}
+    plots = {}
+    figures = {}
+    axes = {}
+    count = 0
+    pyplot.ion() #Interactive mode on
+    configuration = {"bootstrap.servers": server,
+                     "group.id": group,
+                     "default.topic.config" : {
+                         "auto.offset.reset": "earliest"},
+                     "enable.auto.commit": False}
+    consumer = Consumer(configuration)
+    consumer.subscribe([topic])
+    keyboard_input = ERR
+
+    while True:
+        if keyboard_input != ERR:
+            break
+        keyboard_input = screen.getch()
+        message = consumer.consume(num_messages=1, timeout=0.2)
+        if (message is None) or (message == []):
+            redraw(figures)
```

Comment:
Why redraw the figures here? An empty message implies nothing to do. Would `continue` work? I would also recommend a logging message here.

Reply:
Not quite sure what I would be logging here. No new messages in the last one-fifth of a second? Not sure why that would even need to be logged.

Reply:
It's important to log that information. For example, it can tell you if data has stopped flowing. Logging message statistics is critical in debugging and monitoring.

Reply:
Hmmmm. That's important if and only if the data is expected to flow at a rate of more than one message every 0.2 seconds. I could certainly add a command-line parameter for expected data rate, and then add logging to log when the data rate falls below that (maybe another command-line parameter to turn that on and off).

Reply:
I've been thinking about this. To add logging in this particular place, I'd have to add logging at all; this script was never intended as a logging script, so that's adding a new capability. Along with that, I'd need to add new command-line switches; wouldn't it be more straightforward to have an entirely separate logging consumer which logs the messages received from Kafka, and notes periods in which the data has stopped flowing?

That way, this script can maintain its single focus (specifically, on graphing the data so that the user can see it), while any user who wants logging would run the logging consumer as well (perhaps your kafka streaming consumer would even be used for that role?). There's nothing that says you can't have multiple consumers running at once, even on the same computer.